demos/sample_azure_iot/sample_azure_iot.c

```diff
@@ -25,6 +25,11 @@
 /* Crypto helper header. */
 #include "crypto.h"
 
+/* For using the ATECC608 secure element if support is configured */
+#if CONFIG_ESP_TLS_USE_SECURE_ELEMENT && CONFIG_ATCA_MBEDTLS_ECDSA
+    #include "cryptoauthlib.h"
```

Comment:
We'd like to maintain common code to be device-agnostic.
We will provide a different include file with an interface that extracts the registration ID.

Reply:
Sounds good! Do let me know if there is a change you want me to do.

Reply:
Please move any code related to ESP32 to files within demos/projects/espressif/esp32/*. I think this should be moved into `transport_tls_esp32.c`. 

I am still reading about deriving certificates from ATEC608. If I understand correctly, only Trust&Go is guaranteed to always have the same certificate. Our Provisioning Service validates the certificate's key and the Common Name must match the Registration ID.

Since for TrustFlex and TrustCUSTOM the certificate can be changed, the ESP-IDF Azure IoT config should take precendence.

Reply:
> I am still reading about deriving certificates from ATEC608. If I understand correctly, only Trust&Go is guaranteed to always have the same certificate.
Yes, this is always right.

> Since for TrustFlex and TrustCUSTOM the certificate can be changed, the ESP-IDF Azure IoT config should take precendence.
For TrustFLEX, if someone chooses to use the Microchip template certs, then the CN and hence the registration ID can be derived from ATECC608 chip. 
But, I see why accommodating everything can be tricky here.

Reply:
The design I'm thinking of is to either:
1. Use the #defines that we already have
2. Use a common interface to extract information such as registrationID/deviceID, etc (we could, in the future, add hub name, global endpoint, etc)

Reply:
> Use the #defines that we already have

Do we have a #define to allow using any alternative mechanism for private key, cert, etc. if not using the kconfig menu?

If not, how do you feel about checking if the registration ID macro is blank wherever the registration ID is being used? If blank, we call the function that populates the registration ID correctly. This function can then be moved to the `transport_tls_esp32.c` file. 

Pseudocode:
```
...

if registration ID macro is blank

    call the function that calls cryptoauthlib API to populate the registration ID

endif
```

Reply:
> Do we have a #define to allow using any alternative mechanism for private key, cert, etc. if not using the kconfig menu?

We do not, and the sample code is common among many other platforms so we cannot rely on kconfig only.

I propose adding a new `#define USE_HSM` (hardware security module) and a new interface - let's start with a single function to obtain the registration ID.

```
ifdef USE_HSM
   call getRegistrationID() (transport_tls_esp32.c would need to define one specific to reading it from the ATECC chip _if_ the chip is enabled in kconfig).
else
  use the registration ID maco
endif
```

For this application specifically, the developer would need to enable the ATECC chip in kconfig. Ideally, that could also automatically define `USE_HSM`.

This creates a design open to addition of new types of HSMs in the future.

Reply:
Hi @CIPop - I need your feedback.

Based on the comments above, I have altered the implementation such that there are no cryptoauthlib or ESP32 macros/functions in the common code.

For instance, the dynamic registration ID part in `sample_azure_iot_pnp.c` looks like this.
```
#ifdef democonfigUSE_HSM
        /* Redefine the democonfigREGISTRATION_ID macro using registration ID 
        generated dynamically using the HSM */
        
        /* We use a pointer instead of a buffer so that the getRegistrationId
         function can allocate the necessary memory depending on the HSM */
        char *registration_id = NULL;
        ulStatus = getRegistrationId( &registration_id );
        configASSERT( ulStatus == 0);
        #undef democonfigREGISTRATION_ID
        #define democonfigREGISTRATION_ID   registration_id
             
#endif
```

The democonfigUSE_HSM gets defined inside `demo_config.h` depending on whether the use of secure element is enabled as below.
```
/**
 * @brief Defines the macro for HSM usage depending on whether 
 * the support for ATECC608 is enabled in the kconfig menu
 */
#ifdef CONFIG_ESP_TLS_USE_SECURE_ELEMENT
    #if CONFIG_MBEDTLS_ATCA_HW_ECDSA_SIGN & CONFIG_MBEDTLS_ATCA_HW_ECDSA_VERIFY
        #define democonfigUSE_HSM
        uint32_t getRegistrationId( char ** ); // should be defined by provider
    #endif
#endif /* CONFIG_ESP_TLS_USE_SECURE_ELEMENT */

```
One change that may be necessary is that we can no more use sizeof(democonfigREGISTRATION_ID) - 1 - will have to use strlen(democonfigREGISTRATION_ID).
What are your thoughts on this code structure?

**Quick question: Why are we using sizeof(...) - 1 instead of strlen(...) considering these macros are all strings anyway?**

Reply:
Looks good to me! 

> Why are we using sizeof(...) - 1 instead of strlen(...) considering these macros are all strings anyway?

Asked the original devs: It was out of caution to make sure the value isn't determined at runtime, in case the compiler isn't smart enough to collapse `strlen(const)` to the appropriate numerical value. 

I'm OK with either solution:
- use strlen() everywhere 
- create an `#else` branch leaving the old code as-is